From 33f216b16fa22d5ced205c2b0419a881fc11eae8 Mon Sep 17 00:00:00 2001
From: Abhishek <apatgaonkar@habana.ai>
Date: Thu, 22 Aug 2024 08:09:41 +0300
Subject: [PATCH] [FSW-2704] fs1: improve checks in boot flow

This patch improves different checks during new
boot flow for secondary image changes. It checks
for crc failure during manifest validation and
properly compares boot_desc_init return values.

Change-Id: I5745b392427a0a3d4e33c81bf1b9fe0394f0d80e
---

diff --git a/common/boot_flow.c b/common/boot_flow.c
index 2e8a2cb..2fb9abd 100644
--- a/common/boot_flow.c
+++ b/common/boot_flow.c
@@ -161,15 +161,18 @@
 	if (!!boot_rom_global.fw_upd_switch_manifest)
 		curr_manifest_idx = HL_FW_SPI_MANIFEST_1;
 
-	if (boot_rom_verify_manifest(curr_manifest_idx) == BOOT_ROM_SUCCESS)
-		primary_manifest_valid = true;
+	if (boot_rom_boot_desc_init_done(curr_manifest_idx)) {
+		if (boot_rom_verify_manifest(curr_manifest_idx) == BOOT_ROM_SUCCESS)
+			primary_manifest_valid = true;
+	}
 
 	/* TODO: Add check for secondary image disabled */
 
 	curr_manifest_idx = (primary_manifest_idx + 1) % HL_FW_SPI_MANIFEST_MAX_NUM;
-
-	if (boot_rom_verify_manifest(curr_manifest_idx) == BOOT_ROM_SUCCESS)
-		secondary_manifest_valid = true;
+	if (boot_rom_boot_desc_init_done(curr_manifest_idx)) {
+		if (boot_rom_verify_manifest(curr_manifest_idx) == BOOT_ROM_SUCCESS)
+			secondary_manifest_valid = true;
+	}
 }
 
 bool boot_rom_boot_flow_is_manifest_valid(enum boot_rom_manifests rom_manifest_type)
diff --git a/common/common.c b/common/common.c
index e1e1192..fa3cc2a 100644
--- a/common/common.c
+++ b/common/common.c
@@ -77,10 +77,10 @@
 	}
 
 	if (!boot_rom_global.validation_pass) {
-		if (!boot_desc_init(HL_FW_SPI_MANIFEST_0))
+		if (boot_desc_init(HL_FW_SPI_MANIFEST_0) != BOOT_ROM_SUCCESS)
 			boot_rom_status_update(PROGRESS_INIT_SPI_MANIFEST_0_FAIL);
 
-		if (!boot_desc_init(HL_FW_SPI_MANIFEST_1))
+		if (boot_desc_init(HL_FW_SPI_MANIFEST_1) != BOOT_ROM_SUCCESS)
 			boot_rom_status_update(PROGRESS_INIT_SPI_MANIFEST_1_FAIL);
 	}
 
